File: deploy/content_aggregator/aggregate_content.py
from pathlib import Path
import sys
import shutil
import os
import re

class bcolors:
    HEADER = '\033[95m'
    OKBLUE = '\033[94m'
    OKCYAN = '\033[96m'
    OKGREEN = '\033[92m'
    WARNING = '\033[93m'
    FAIL = '\033[91m'
    ENDC = '\033[0m'
    BOLD = '\033[1m'
    UNDERLINE = '\033[4m'

p = Path(sys.path[0]+"/../../public/content")


# Read content dirs and assemble shelves
content_files = list(p.glob("./**/*"))

shelves = {}

for path in content_files:
    content_url = str(path).replace(sys.path[0]+"/../../public/", "")

    shelf = content_url.split("/")[-2]

    if (shelf == "content"): continue

    article_title = (".".join(content_url.split(".")[:-1])).split("/")[-1]
    print(f"{bcolors.OKBLUE}{shelf}{bcolors.ENDC}: {article_title}")

    filetype = content_url.split(".")[-1] # Don't lie to me!

    # TODO: tag parsing on md files to sort by date

    article_html = f'''
<article
    class="book"
    draggable="true"
    ondragstart="book_dragstart(event)"
    ondragend="book_dragend(event)"
    onpointerup="book_pointerup(event)"
    data-contenturl="{content_url}"
    data-title="{article_title}"
>
    <h2>{article_title}</h2>
</article>
    '''

    if shelf not in shelves:
        shelves[shelf] = ""

    shelves[shelf] += article_html

#print(shelves)

print()

# Copy index (so we dont lose it!)
try:
    os.mkdir(f"{sys.path[0]}/../../backup")
except FileExistsError: 
    print("Backup folder already exists")

shutil.copy(f"{sys.path[0]}/../../index.html", f"{sys.path[0]}/../../backup/index.html")

# Write shelves into index html in designated place
data = ""
with open(f"{sys.path[0]}/../../index.html", "r") as indexHtml:
    data = indexHtml.read()

shelf_template = '''<!-- SHELF: shelftitle -->
shelfcontent
<!-- STOP SHELF: shelftitle -->
'''

shelf_regex = r'<!-- SHELF: shelftitle -->(.|\n)*<!-- STOP SHELF: shelftitle -->'

seen_shelves = []

print()

for shelf in shelves:
    shelf_html = shelves[shelf]

    if not re.search(shelf_regex.replace("shelftitle", shelf), data):
        print(f"{bcolors.FAIL}Hey! No {shelf}{bcolors.ENDC} shelf")
    else:
        data = re.sub(
                shelf_regex.replace("shelftitle", shelf),
                shelf_template.replace("shelftitle", shelf).replace("shelfcontent", shelf_html),
                data,
                flags=re.M)
    

        print(f"Seen shelf: {shelf}")
        seen_shelves.append(shelf)

with open(f"{sys.path[0]}/../../index.html", "w") as writeIndex:
    writeIndex.write(data)
    

File: deploy/content_aggregator/aggregate_content.js
import * as fs from "node:fs/promises";
import * as path from "node:path";
import { parse_md } from "./parse_md.js";
import { shelf_regex, bcolors } from "./util.js";

/** @typedef {object} article
 * @property {string} title
 * @property {string} filetype
 * @property {string} content_url
 * @property {number} timestamp
 * @property {string} html
 */

const shelf_template = `<!-- SHELF: shelftitle -->
shelfcontent
<!-- STOP SHELF: shelftitle -->`;

let shelves = {};

const diary_sorter = (articles) => {
  let series_names = [];
  for (const article of articles) {
    let series_name = article.title.split(" ")[0].toLowerCase();
    if (series_name === "the") {
      series_name += " " + article.title.split(" ")[1].toLowerCase();
    }
    if (!series_names.includes(series_name)) {
      series_names.push(series_name);
    }
  }

  let sorted_shelf = [];
  for (const series_name of series_names) {
    let sorted_series = articles.filter((a) =>
      a.title.toLowerCase().startsWith(series_name),
    );

    console.log(sorted_series.map((a) => a.title));

    sorted_series = sorted_series.sort(
      (a, b) => (b.timestamp ?? 0) - (a.timestamp ?? 0),
    );

    sorted_shelf = sorted_shelf.concat(sorted_series);
  }

  return sorted_shelf;
};

const ignored_folders = ["public", "content"];

const aggregate_content = async () => {
  console.log(import.meta.dirname);
  const read = await fs.glob(import.meta.dirname + "/../../public/**/*");

  for await (const f of read) {
    const filename = path.basename(f);
    const dir = path.dirname(f);

    const shelf = dir.split("/")[dir.split("/").length - 1];
    const title = filename
      .split(".")
      .filter((sp, idx) => idx !== filename.split(".").length - 1)
      .join(".");
    const filetype = filename.split(".")[filename.split(".").length - 1];

    if (ignored_folders.includes(shelf)) {
      continue;
    }
    console.log(shelf, title, filetype);

    if (!shelves[shelf]) {
      shelves[shelf] = [];
    }

    // Harvest articles for relevant data
    let article = {
      title,
      filetype,
      content_url: dir.replace(/^.+?content/, "content") + `/${filename}`,
      date: null,
      html: "",
    };

    switch (filetype) {
      case "md": {
        // TODO: tag articles with tags
        await parse_md(f, article);
        break;
      }
      default: {
      }
    }

    shelves[shelf].push(article);
  }

  // Read index
  // TODO: make copy of index
  let index = await fs.readFile(`${import.meta.dirname}/../../index.html`, {
    encoding: "utf8",
  });

  for (const shelfName of Object.keys(shelves)) {
    // Sort articles by date
    let sortedShelf = shelves[shelfName].sort(
      (a, b) => (b.timestamp ?? 0) - (a.timestamp ?? 0),
    );

    // Special sort for diary
    if (shelfName === "dagloopyrpgdiary") {
      sortedShelf = diary_sorter(shelves[shelfName]);
      console.log(sortedShelf.map((a) => a.title));
    }

    console.log(`Pushing ${bcolors.OKBLUE}'${shelfName}'${bcolors.ENDC}`);

    const new_article_html = shelf_template
      .replaceAll("shelftitle", shelfName)
      .replace("shelfcontent", sortedShelf.map((a) => a.html).join("\n"));

    const r = shelf_regex.replaceAll("shelftitle", shelfName);
    index = index.replace(new RegExp(r), new_article_html);
  }

  if (process.argv.includes("--deploy")) {
    fs.writeFile(`${import.meta.dirname}/../../index.html`, index, {
      encoding: "utf8",
    });
  } else {
    console.log(
      `${bcolors.OKBLUE}WOULD push out ${Object.entries(shelves).reduce((total, [sName, s]) => total + s.length, 0)} articles. To do this for real, re-run with ${bcolors.WARNING}--deploy`,
    );
  }
};

aggregate_content();
